[김선호] 인가(Authorization) 리뷰 요청 드립니다.

haero77 opened the pull request:
시큐리티 코드를 뜯어보면서 공부하느라 시간을 조금 많이 썼네요.. 🥲

놓친 요구사항이나 구조적으로 아쉬운 부분 있으면 얼마든지 말씀 부탁드려요!! 

Q. 미션을 구현하다보니 RequestAuthorizationManager에서 Genric Raw Use 관련해서 경고가 뜨던데 이 부분을 어떻게 핸들링하기가 어렵더라고요. 시큐리티 코드를 보니까 `List<RequestMatcherEntry<AuthorizationManager<RequestAuthorizationContext>>>` 같은 형태를 쓰긴 하던데 현재 수준에서는 Genric Raw Use를 어떻게 피할지 감이 잘 안 잡히네요. 😅 힌트 주시면 감사하겠습니다!

그럼 이번 리뷰 잘 부탁드립니다..! 🙇‍♂️

haero77 committed:
code polishing

haero77 committed:
test: Form Login 로직을 test fixture로 추출

haero77 committed:
(실습) GET /members/me 엔드포인트 추가 및 테스트 작성

haero77 committed:
fix: FilterChainProxy가 필터 체인에 등록되어 SecurityFilterChain이 중복으로 실행되는 버그 수정

- FilterChainProxy는 빈으로 등록 안 하고, DelegatingFilterProxy만 빈으로 등록하여 서블릿 필터 체인에는 FilterChainProxy이 등록되지 않게 함으로써 해결

haero77 committed:
(실습) 엔드포인트에 따라 인가를 검사하는 AuthorizationFilter 구현

haero77 committed:
(step1) 요청 자원의 권한을 확인하는 AuthorizationManager, 인가 결과를 담는 AuthorizationDecision 정의

haero77 committed:
(step1) 요청 URI에 따라 인가 여부를 확인하는 로직을 RequestAuthorizationManager에게 위임하도록 변경

haero77 committed:
(step1) Secured 애노테이션을 사용한 인가 로직을 SecuredAuthorizationManager가 담당하도록 변경

haero77 committed:
(step2) 요청 일치 여부 확인을 위한 RequestMatcher 정의 및 AnyRequestMatcher, MvcRequestMatcher 구현

haero77 committed:
(step2) RequestMatcher 와 레퍼런스 타입을 함께 관리하는 RequestMatcherEntry 추가

haero77 committed:
(step2) 인증된 사용자는 모두 인가 처리하는 AuthenticatedAuthorizationManager 추가

haero77 committed:
(step2) Authentication의 권한을 확인하고 인가하는 AuthorityAuthorizationManager 구현

haero77 committed:
(step2) 모든 요청에 대해 인가를 하지 않는 DenyAllAuthorizationManager 구현

haero77 committed:
(step2) 모든 요청에 대해 인가를 하는 PermitAllAuthorizationManager 구현

haero77 committed:
(step2) 요청 패턴에 따라 적절한 인가 관리자를 호출하여 인가하도록 인가 로직 위임

parkeeseul commented on src/main/java/nextstep/app/ui/MemberController.java:
```diff
@@ -24,6 +27,19 @@ public ResponseEntity<List<Member>> list() {
+        }
+
+        return ResponseEntity.ok(memberOpt.get());
+    }
```
orElseThrow()를 사용하면 더 간결하고 처리할 수 있겠네요 😄 
```suggestion
        Member member = memberRepository.findByEmail(email)
                .orElseThrow(() -> new IllegalStateException("회원을 찾을 수 없습니다. (email=%s)".formatted(email)));

        return ResponseEntity.ok(member);
    }
```

haero77 replied:
[dcc8acc](https://github.com/next-step/spring-security-authorization/pull/27/commits/dcc8acc913ffd343d8ef45a7932d93f6ba8c9240) 커밋에 해당 내용 반영해봤습니다 :)

---

별개로 조금 다른 질문이긴 하지만, 다음 두 가지 내용이 궁금해서 질문드려봅니다..!

1. _위 `MemberController::getMember` 로직에서 `IllegalStateException`과 `IllegalArgumentException` 중 무엇이 더 적절할까요?_

두 예외 중 무엇을 발생시킬 지 고민했는데, 이미 인증 필터에서 인증된 상태의 유저를 찾았고 이 상태에서 username과 일치하는 member가 없다는 것은 클라이언트의 잘못(IllegalArgumentException)이라기 보다는 서버 상태의 이상을 의미하기 때문에 IllegalStateException으로 선언했습니다. 리뷰어님이라면 어떤 예외를 사용하셨을지 궁금합니다..!

2. _컨트롤러에서 SecurityContext를 직접 접근하는게 괜찮을까요?_

컨트롤러는 Spring MVC의 영역이고, 인증이 다 끝난 상태에서 수행되는 비즈니스 로직인데 여기서 SecurityContext를 접근해도 괜찮은 걸까라는 고민이 되더군요. 또, 실무 코드라면 memberRepository.findByEmail등의 로직을 매번 작성해야하는 것이 번거로워 `@AuthenticatedMember`라는 애노테이션을 만들고 MethodArgumentResolver로 Member객체를 주입해줄 것 같긴 한데 여기에 대한 리뷰어님의 의견이 궁금합니다 :-)

parkeeseul replied:
> 1. 위 MemberController::getMember 로직에서 IllegalStateException과 IllegalArgumentException 중 무엇이 더 적절할까요?

A. email이 조회되지 않는다면, 말씀하신 것 처럼 애플리케이션의 상태가 예상과 다르게 흘러갔다고 볼 수 있을 것 같아요.
저도 선호님 처럼 IllegalStateException 이 더 적절하다고 생각합니다.
 다만 실무에서는 CustomException을 정의하여 보다 명확한 예외를 던지는 방식을 선택할 것 같네요.  
 CustomException을 통해 예외의 의미를 명확하게 전달하고 일관된 방식으로 예외를 처리하는 방식을 선호합니다.😄 

> 2. 컨트롤러에서 SecurityContext를 직접 접근하는 것이 적절한가?

A.저도 실무에서는 선호님께서 말씀해 주신 것처럼 커스텀 애노테이션을 만들어 사용하는 방식을 선호합니다. 😄
컨트롤러는 보안 관련 로직을 다루기보다는, 요청을 처리하는 역할에 집중하는 것이 좋다고 생각해요.
또한, 인증 방식이 변경될 때 컨트롤러를 수정할 필요 없이 ArgumentResolver만 변경하면 되므로, 유지보수적인 측면에서도 더 좋은 방식이라고 생각합니다. 👍

parkeeseul commented on src/main/java/nextstep/security/authorization/SecuredAuthorizationManager.java:
```diff
@@ -0,0 +1,26 @@
+        Method method = invocation.getMethod();
+
+        if (method.isAnnotationPresent(Secured.class)) {
+            Secured secured = method.getAnnotation(Secured.class);
```
Spring AOP 환경에서는 메서드가 프록시 객체로 감싸질 수 있기 때문에, 프록시된 메서드를 조회할 경우 실제 구현체의 메서드에서 선언된 `@Secured` 어노테이션을 찾지 못할 가능성이 있습니다.
method.getAnnotation 와 AopUtils.getMostSpecificMethod() 어떤 차이가 있을까요? 😄

haero77 replied:
`method.getAnnotation`의 경우, 단순히 호출된 메서드를 찾기 때문에 (AOP로 인해 프록시 객체가 생성되어) 프록시 메서드가 조회되는 경우 애노테이션 정보를 못 찾을 수 있습니다. (프록시 메서드에는 애노테이션 정보가 없기 때문)

반면에 `AopUtils.getMostSpecificMethod()` 같은 경우 프록시 객체가 아닌 실제 구현체의 메서드를 찾기 때문에 애노테이션 정보를 확실하게 찾을 수 있겠군요...! 

---

해당 내용 [52bd8f5](https://github.com/next-step/spring-security-authorization/pull/27/commits/52bd8f5662d68cf3993a5cc841366b2f8a75bcde) 커밋에 반영해봤습니다!

parkeeseul commented on src/main/java/nextstep/security/util/matcher/AnyRequestMatcher.java:
```diff
@@ -0,0 +1,11 @@
+
+import jakarta.servlet.http.HttpServletRequest;
+
+public class AnyRequestMatcher implements RequestMatcher {
```
AnyRequestMatcher는 동일한 동작을 수행하는 불변 객체이므로 싱글톤 패턴 적용을 해봐도 좋을 것 같아요 😄 
또한, 외부에서 인스턴스 생성을 방지하도록 private 생성자를 추가하는 것도 좋겠네요😄

``` java
public static final AnyRequestMatcher INSTANCE = new AnyRequestMatcher();

private AnyRequestMatcher() {
}
```

haero77 replied:
앞으로도 Spring Bean으로 등록되지는 않지만, 애플리케이션 내에서 항상 동일한 동작을 보장하는 경우 싱글톤으로 선언할 수 있을지 신경써봐야겠네요. 인사이트 감사합니다 :)

---

해당 내용 [601ad5d](https://github.com/next-step/spring-security-authorization/pull/27/commits/601ad5dfed052a690d0435b939fb302c5188573a) 커밋에 반영해봤습니다.

parkeeseul commented on src/test/java/nextstep/app/FormLoginTest.java:
```diff
@@ -85,35 +94,94 @@ void admin_login_after_members() throws Exception {
+                .andExpect(status().isForbidden());
+    }
+
+    @DisplayName("인증된 사용자는 자신의 정보를 조회할 수 있다.")
```
크으 테스트 코드 👍

parkeeseul commented on src/main/java/nextstep/security/authorization/RequestAuthorizationManager.java:
```diff
@@ -0,0 +1,40 @@
+ * - 요청에 대해서 적합한 권한이 있는지만 판단.
+ * - 인가가 되지 않았을 경우 예외를 발생시키지는 않는다. (예외는 AuthorizationFilter에서 던짐)
+ */
+public class RequestAuthorizationManager implements AuthorizationManager<HttpServletRequest> {
```
> Q. 미션을 구현하다보니 RequestAuthorizationManager에서 Genric Raw Use 관련해서 경고가 뜨던데 이 부분을 어떻게 핸들링하기가 어렵더라고요. 시큐리티 코드를 보니까 List<RequestMatcherEntry<AuthorizationManager<RequestAuthorizationContext>>> 같은 형태를 쓰긴 하던데 현재 수준에서는 Genric Raw Use를 어떻게 피할지 감이 잘 안 잡히네요. 😅 힌트 주시면 감사하겠습니다!

A. 제네릭 타입을 명확히 지정하면 더 안전한 코드가 될 것 같아요! 
아래와 같이 `AuthorizationManager<HttpServletRequest>`를 지정하면  
IDE 경고도 사라지고, 타입 안정성이 더 높아집니다.

``` java
private final List<RequestMatcherEntry<AuthorizationManager<HttpServletRequest>>> mappings;
```

haero77 replied:
인스턴스 필드 `mappings`에 할당되는 AuthorizationManager 구현체들이 전부 HttpServletRequest를 지원하는 것은 아니라서  , 

```java
private final List<RequestMatcherEntry<AuthorizationManager<HttpServletRequest>>> mappings;
```

처럼 사용하게 될 경우 컴파일 에러가 발생하게 되더라고요 😅

<img width="894" alt="image" src="https://github.com/user-attachments/assets/a11d6d15-cbdf-4c29-b063-5650ce073181" />

지금 수준에서 타입 안정성을 챙길 수 있는 방법이 있을지 더 고민해봐야겠습니다..!

parkeeseul replied:
다음과 같이 해결할 수 있을 것 같네요 😄 
RequestAuthorizationManager`는 `AuthorizationManager<HttpServletRequest>`를 기반으로 동작하므로 아래처럼 진행해도 문제 없을 것 같아요 😄 
![스크린샷 2025-02-25 오전 12 04 36](https://github.com/user-attachments/assets/f7532ac3-6594-46bb-8574-a3115a5b35be)
![스크린샷 2025-02-25 오전 12 04 50](https://github.com/user-attachments/assets/a50ae224-7c09-4d8d-ae49-b321b7018f2b)

parkeeseul reviewed (changes requested):
안녕하세요! 인가 미션을 함께 하게 된 박이슬 입니다 😄 
미션 잘 진행해 주셨네요! 👍 
몇 가지 코멘트 남겨두었으니 확인 부탁드립니다
미션 화이팅 입니다 🔥

haero77 committed:
(feedback) Optional::orElseThrow를 이용한 예외 처리 로직 단순화

- resolve https://github.com/next-step/spring-security-authorization/pull/27#discussion_r1957261715

haero77 committed:
(feedback) 항상 동일한 동작을 보장하는 AnyRequestMatcher를 싱글톤으로 리팩토링

- resolve https://github.com/next-step/spring-security-authorization/pull/27#discussion_r1957261715

haero77 committed:
(feedback) AopUtils.getMostSpecificMethod를 사용하여 AOP로 인한 프록시 객체가 생성될 경우에도 Secured 애노테이션을 찾을 수 있도록 개선

- resolve https://github.com/next-step/spring-security-authorization/pull/27#discussion_r1957299663

haero77 committed:
docs: add Questions.md

parkeeseul reviewed (approved):
피드백 잘 반영해 주셨네요 👍 고생 많으셨습니다!
소소한 코멘트 남겨두었으니 확인 부탁드리고 이만 머지하도록 하겠습니다.
해당 미션은 선택 미션도 있는데요, 혹시 여유가 되신다면 진행하셔도 좋을 것 같습니다.
남은 미션도 화이팅 입니다 🔥
